storage/mysql/log_storage.go

```diff
@@ -322,27 +331,31 @@ func (t *logTreeTX) QueueLeaves(leaves []*trillian.LogLeaf, queueTimestamp time.
 		insertSQL = insertUnsequencedLeafSQLNoDuplicates
 	}
 
-	// Insert in order of the hash values in the leaves.
-	orderedLeaves := make([]*trillian.LogLeaf, len(leaves))
-	copy(orderedLeaves, leaves)
-	sort.Sort(byLeafIdentityHash(orderedLeaves))
+	// Insert in order of the hash values in the leaves, but track original position for return value.
+	orderedLeaves := make([]leafAndPosition, len(leaves))
+	for i, leaf := range leaves {
+		orderedLeaves[i] = leafAndPosition{leaf: leaf, idx: i}
+	}
+	sort.Sort(byLeafIdentityHashWithPosition(orderedLeaves))
+	existingCount := 0
+	existingLeaves := make([]*trillian.LogLeaf, len(leaves))
 
-	for i, leaf := range orderedLeaves {
+	for i, leafPos := range orderedLeaves {
+		leaf := leafPos.leaf
 		// Create the unsequenced leaf data entry. We don't use INSERT IGNORE because this
 		// can suppress errors unrelated to key collisions. We don't use REPLACE because
 		// if there's ever a hash collision it will do the wrong thing and it also
 		// causes a DELETE / INSERT, which is undesirable.
 		_, err := t.tx.Exec(insertSQL, t.treeID, leaf.LeafIdentityHash, leaf.LeafValue, leaf.ExtraData)
+		if err != nil && strings.Contains(err.Error(), "Duplicate entry") {
```

Comment:
This code is in a mysql specific package so I don't think we should worry about that yet though it should test the error code at some point.

I thought we were going to do a SELECT ... FOR UPDATE pass to find dups first.